File: torchao/experimental/kernels/cpu/linear/examples/torch_custom_op/torch_custom_op.cpp
// Copyright (c) Meta Platforms, Inc. and affiliates.
// All rights reserved.
//
// This source code is licensed under the license found in the
// LICENSE file in the root directory of this source tree.

#include <torch/library.h>
#include <torch/script.h>
#include <torch/torch.h>
#include <torchao/experimental/kernels/cpu/linear/channelwise_8bit_activation_groupwise_lowbit_weight.h>
#include <torchao/experimental/kernels/cpu/parallel.h>

template <int weight_nbit>
at::Tensor pack_weights_without_zeros_cpu(
    const at::Tensor& weight_qvals,
    const at::Tensor& weight_scales,
    // TODO(T200095131): convert to int64_t when supported by AOTI
    // group_size is a meta tensor with size (group_size)
    const at::Tensor& group_size_tensor) {
  int64_t group_size = group_size_tensor.size(0);

  TORCH_CHECK(
      weight_qvals.dtype() == torch::kInt8, "weight_qvals must be int8");
  TORCH_CHECK(weight_qvals.dim() == 2, "weight_qvals must be 2D");

  // In PyTorch, weights are nxk in row-major format (with activations being
  // right-multiplied).
  // In kernel, activations are left-multiplied by kxn transposed
  // weights in column-major format.
  // Note the underlying data is the same in both cases
  int n = weight_qvals.size(0);
  int k = weight_qvals.size(1);

  TORCH_CHECK(
      weight_scales.dtype() == torch::kFloat32,
      "weight_scales must be float32");
  TORCH_CHECK(weight_scales.dim() == 1, "weight_scales must be 1D");
  TORCH_CHECK(
      weight_scales.size(0) == ((n * k) / group_size),
      "expected 1 scale per group");

  using namespace torchao::operators::cpu::linear::
      channelwise_8bit_activation_groupwise_lowbit_weight;

  auto ukernel_config = get_ukernel_config<
      weight_nbit,
      false /*has_weight_zeros*/,
      false /*has_bias*/,
      false /*has_clamp*/>();
  auto pack_weight_tiling_params = get_default_pack_weight_data_tiling_params(
      ukernel_config, n, /*target_panels_per_thread=*/1);

  auto packed_weight_data_size =
      get_packed_weight_data_size(ukernel_config, n, k, group_size);
  at::Tensor packed_weights =
      torch::empty({packed_weight_data_size}, torch::kInt8);
  pack_weight_data_operator(
      ukernel_config,
      pack_weight_tiling_params,
      packed_weights.data_ptr<int8_t>(),
      n,
      k,
      group_size,
      weight_qvals.const_data_ptr<int8_t>(),
      weight_scales.const_data_ptr<float>(),
      /*weight_zeros=*/nullptr);

  return packed_weights;
}

template <int weight_nbit>
at::Tensor pack_weights_with_zeros_cpu(
    const at::Tensor& weight_qvals,
    const at::Tensor& weight_scales,
    const at::Tensor& weight_zeros,
    // TODO(T200095131): convert to int64_t when supported by AOTI
    // group_size is a meta tensor with size (group_size)
    const at::Tensor& group_size_tensor) {
  int64_t group_size = group_size_tensor.size(0);

  TORCH_CHECK(
      weight_qvals.dtype() == torch::kInt8, "weight_qvals must be int8");
  TORCH_CHECK(weight_qvals.dim() == 2, "weight_qvals must be 2D");

  // In PyTorch, weights are nxk in row-major format (with activations being
  // right-multiplied).
  // In kernel, activations are left-multiplied by kxn transposed
  // weights in column-major format.
  // Note the underlying data is the same in both cases
  int n = weight_qvals.size(0);
  int k = weight_qvals.size(1);

  TORCH_CHECK(
      weight_scales.dtype() == torch::kFloat32,
      "weight_scales must be float32");
  TORCH_CHECK(weight_scales.dim() == 1, "weight_scales must be 1D");
  TORCH_CHECK(
      weight_scales.size(0) == ((n * k) / group_size),
      "expected 1 scale per group");
  TORCH_CHECK(
      weight_zeros.dtype() == torch::kInt8, "weight_zeros must be int8");
  TORCH_CHECK(weight_zeros.dim() == 1, "weight_zeros must be 1D");
  TORCH_CHECK(
      weight_zeros.size(0) == ((n * k) / group_size),
      "expected 1 zero per group");

  using namespace torchao::operators::cpu::linear::
      channelwise_8bit_activation_groupwise_lowbit_weight;

  auto ukernel_config = get_ukernel_config<
      weight_nbit,
      true /*has_weight_zeros*/,
      false /*has_bias*/,
      false /*has_clamp*/>();
  auto pack_weight_tiling_params = get_default_pack_weight_data_tiling_params(
      ukernel_config, n, /*target_panels_per_thread=*/1);

  auto packed_weight_data_size =
      get_packed_weight_data_size(ukernel_config, n, k, group_size);
  at::Tensor packed_weights =
      torch::empty({packed_weight_data_size}, torch::kInt8);
  pack_weight_data_operator(
      ukernel_config,
      pack_weight_tiling_params,
      packed_weights.data_ptr<int8_t>(),
      n,
      k,
      group_size,
      weight_qvals.const_data_ptr<int8_t>(),
      weight_scales.const_data_ptr<float>(),
      weight_zeros.const_data_ptr<int8_t>());

  return packed_weights;
}

template <int weight_nbit>
at::Tensor pack_weights_without_zeros_meta(
    const at::Tensor& weight_qvals,
    const at::Tensor& weight_scales,
    // TODO(T200095131): convert to int64_t when supported by AOTI
    // group_size is a meta tensor with size (group_size)
    const at::Tensor& group_size_tensor) {
  int64_t group_size = group_size_tensor.size(0);

  int n = weight_qvals.size(0);
  int k = weight_qvals.size(1);

  using namespace torchao::operators::cpu::linear::
      channelwise_8bit_activation_groupwise_lowbit_weight;

  auto ukernel_config = get_ukernel_config<
      weight_nbit,
      false /*has_weight_zeros*/,
      false /*has_bias*/,
      false /*has_clamp*/>();

  auto packed_weight_data_size =
      get_packed_weight_data_size(ukernel_config, n, k, group_size);
  return torch::empty({packed_weight_data_size}).to("meta");
}

template <int weight_nbit>
at::Tensor pack_weights_with_zeros_meta(
    const at::Tensor& weight_qvals,
    const at::Tensor& weight_scales,
    const at::Tensor& weight_zeros,
    // TODO(T200095131): convert to int64_t when supported by AOTI
    // group_size is a meta tensor with size (group_size)
    const at::Tensor& group_size_tensor) {
  int64_t group_size = group_size_tensor.size(0);

  int n = weight_qvals.size(0);
  int k = weight_qvals.size(1);

  using namespace torchao::operators::cpu::linear::
      channelwise_8bit_activation_groupwise_lowbit_weight;

  auto ukernel_config = get_ukernel_config<
      weight_nbit,
      true /*has_weight_zeros*/,
      false /*has_bias*/,
      false /*has_clamp*/>();

  auto packed_weight_data_size =
      get_packed_weight_data_size(ukernel_config, n, k, group_size);
  return torch::empty({packed_weight_data_size}).to("meta");
}

template <int weight_nbit, bool has_weight_zeros>
at::Tensor linear_cpu(
    const at::Tensor& packed_weights,
    // TODO(T200095131): convert n_tensor, k_tensor, group_size_tensor to
    // int64_t when supported by AOTI Currently they are meta tensors with size
    // equal to the int they wrap
    const at::Tensor& n_tensor,
    const at::Tensor& k_tensor,
    const at::Tensor& group_size_tensor,
    const at::Tensor& activations) {
  int n = n_tensor.size(0);
  int k = k_tensor.size(0);
  int group_size = group_size_tensor.size(0);

  TORCH_CHECK(
      activations.dtype() == torch::kFloat32, "activations must be float32");
  TORCH_CHECK(activations.dim() == 2, "activations must be 2D");
  int m = activations.size(0);
  int k_ = activations.size(1);
  TORCH_CHECK(k == k_, "activation shape is incompatible with packed weights.");

  using namespace torchao::operators::cpu::linear::
      channelwise_8bit_activation_groupwise_lowbit_weight;

  auto ukernel_config = get_ukernel_config<
      weight_nbit,
      has_weight_zeros /*has_weight_zeros*/,
      false /*has_bias*/,
      false /*has_clamp*/>();
  auto linear_tiling_params = get_default_linear_tiling_params(
      ukernel_config,
      m,
      n,
      /*target_tiles_per_thread=*/5);
  auto linear_scheduling_policy =
      LinearTileSchedulingPolicy::single_mc_parallel_nc;

  auto activation_data_buffer_size = get_activation_data_buffer_size(
      ukernel_config,
      linear_tiling_params,
      linear_scheduling_policy,
      m,
      k,
      group_size);
  std::vector<char> activation_data_buffer(activation_data_buffer_size);

  at::Tensor output_tensor = torch::empty({m, n}, torch::kFloat32);
  linear_operator(
      ukernel_config,
      linear_tiling_params,
      linear_scheduling_policy,
      activation_data_buffer.data(),
      output_tensor.data_ptr<float>(),
      m,
      n,
      k,
      group_size,
      packed_weights.const_data_ptr<int8_t>(),
      activations.const_data_ptr<float>(),
      /*bias=*/nullptr,
      // Clamp parameters are ignored because config is created from
      // has_clamp = false
      /*clamp_min=*/0.0,
      /*clamp_max=*/0.0);

  return output_tensor;
}

template <int weight_nbit, bool has_weight_zeros>
at::Tensor linear_meta(
    const at::Tensor& packed_weights,
    // TODO(T200095131): convert n_tensor, k_tensor, group_size_tensor to
    // int64_t when supported by AOTI
    // Currently they are meta tensors with size equal to the int they wrap
    const at::Tensor& n_tensor,
    const at::Tensor& k_tensor,
    const at::Tensor& group_size_tensor,
    const at::Tensor& activations) {
  int n = n_tensor.size(0);
  int k = k_tensor.size(0);

  int m = activations.size(0);
  int k_ = activations.size(1);
  TORCH_CHECK(k == k_, "activation shape is incompatible with packed weights.");
  return torch::empty({m, n}).to("meta");
}

TORCH_LIBRARY(torchao, m) {
  // Pack weights without zeros
  m.def(
      "_pack_weights_a8sz_w2s(Tensor weight_qvals, Tensor weight_scales, Tensor group_size) -> Tensor");
  m.def(
      "_pack_weights_a8sz_w3s(Tensor weight_qvals, Tensor weight_scales, Tensor group_size) -> Tensor");
  m.def(
      "_pack_weights_a8sz_w4s(Tensor weight_qvals, Tensor weight_scales, Tensor group_size) -> Tensor");
  m.def(
      "_pack_weights_a8sz_w5s(Tensor weight_qvals, Tensor weight_scales, Tensor group_size) -> Tensor");
  // Pack weights with zeros
  m.def(
      "_pack_weights_a8sz_w2sz(Tensor weight_qvals, Tensor weight_scales, Tensor weight_zeros, Tensor group_size) -> Tensor");
  m.def(
      "_pack_weights_a8sz_w3sz(Tensor weight_qvals, Tensor weight_scales, Tensor weight_zeros, Tensor group_size) -> Tensor");
  m.def(
      "_pack_weights_a8sz_w4sz(Tensor weight_qvals, Tensor weight_scales, Tensor weight_zeros, Tensor group_size) -> Tensor");
  m.def(
      "_pack_weights_a8sz_w5sz(Tensor weight_qvals, Tensor weight_scales, Tensor weight_zeros, Tensor group_size) -> Tensor");
  // Linear weights without zeros
  m.def(
      "_linear_a8sz_w2s(Tensor packed_weights, Tensor n, Tensor k, Tensor group_size, Tensor activations) -> Tensor");
  m.def(
      "_linear_a8sz_w3s(Tensor packed_weights, Tensor n, Tensor k, Tensor group_size, Tensor activations) -> Tensor");
  m.def(
      "_linear_a8sz_w4s(Tensor packed_weights, Tensor n, Tensor k, Tensor group_size, Tensor activations) -> Tensor");
  m.def(
      "_linear_a8sz_w5s(Tensor packed_weights, Tensor n, Tensor k, Tensor group_size, Tensor activations) -> Tensor");
  // Linear weights with zeros
  m.def(
      "_linear_a8sz_w2sz(Tensor packed_weights, Tensor n, Tensor k, Tensor group_size, Tensor activations) -> Tensor");
  m.def(
      "_linear_a8sz_w3sz(Tensor packed_weights, Tensor n, Tensor k, Tensor group_size, Tensor activations) -> Tensor");
  m.def(
      "_linear_a8sz_w4sz(Tensor packed_weights, Tensor n, Tensor k, Tensor group_size, Tensor activations) -> Tensor");
  m.def(
      "_linear_a8sz_w5sz(Tensor packed_weights, Tensor n, Tensor k, Tensor group_size, Tensor activations) -> Tensor");
}

TORCH_LIBRARY_IMPL(torchao, CPU, m) {
  m.impl("_pack_weights_a8sz_w2s", &pack_weights_without_zeros_cpu<2>);
  m.impl("_pack_weights_a8sz_w3s", &pack_weights_without_zeros_cpu<3>);
  m.impl("_pack_weights_a8sz_w4s", &pack_weights_without_zeros_cpu<4>);
  m.impl("_pack_weights_a8sz_w5s", &pack_weights_without_zeros_cpu<5>);
  m.impl("_pack_weights_a8sz_w2sz", &pack_weights_with_zeros_cpu<2>);
  m.impl("_pack_weights_a8sz_w3sz", &pack_weights_with_zeros_cpu<3>);
  m.impl("_pack_weights_a8sz_w4sz", &pack_weights_with_zeros_cpu<4>);
  m.impl("_pack_weights_a8sz_w5sz", &pack_weights_with_zeros_cpu<5>);
  m.impl("_linear_a8sz_w2s", &linear_cpu<2, false>);
  m.impl("_linear_a8sz_w3s", &linear_cpu<3, false>);
  m.impl("_linear_a8sz_w4s", &linear_cpu<4, false>);
  m.impl("_linear_a8sz_w5s", &linear_cpu<5, false>);
  m.impl("_linear_a8sz_w2sz", &linear_cpu<2, true>);
  m.impl("_linear_a8sz_w3sz", &linear_cpu<3, true>);
  m.impl("_linear_a8sz_w4sz", &linear_cpu<4, true>);
  m.impl("_linear_a8sz_w5sz", &linear_cpu<5, true>);
}

TORCH_LIBRARY_IMPL(torchao, Meta, m) {
  m.impl("_pack_weights_a8sz_w2s", &pack_weights_without_zeros_meta<2>);
  m.impl("_pack_weights_a8sz_w3s", &pack_weights_without_zeros_meta<3>);
  m.impl("_pack_weights_a8sz_w4s", &pack_weights_without_zeros_meta<4>);
  m.impl("_pack_weights_a8sz_w5s", &pack_weights_without_zeros_meta<5>);
  m.impl("_pack_weights_a8sz_w2sz", &pack_weights_with_zeros_meta<2>);
  m.impl("_pack_weights_a8sz_w3sz", &pack_weights_with_zeros_meta<3>);
  m.impl("_pack_weights_a8sz_w4sz", &pack_weights_with_zeros_meta<4>);
  m.impl("_pack_weights_a8sz_w5sz", &pack_weights_with_zeros_meta<5>);
  m.impl("_linear_a8sz_w2s", &linear_meta<2, false>);
  m.impl("_linear_a8sz_w3s", &linear_meta<3, false>);
  m.impl("_linear_a8sz_w4s", &linear_meta<4, false>);
  m.impl("_linear_a8sz_w5s", &linear_meta<5, false>);
  m.impl("_linear_a8sz_w2sz", &linear_meta<2, true>);
  m.impl("_linear_a8sz_w3sz", &linear_meta<3, true>);
  m.impl("_linear_a8sz_w4sz", &linear_meta<4, true>);
  m.impl("_linear_a8sz_w5sz", &linear_meta<5, true>);
}

File: torchao/experimental/kernels/cpu/linear/examples/torch_custom_op/torch_custom_op.py
# Copyright (c) Meta Platforms, Inc. and affiliates.
# All rights reserved.
#
# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.

from typing import Optional

import torch
import torch.nn as nn

torch.ops.load_library(
    "/tmp/cmake-out/torch_ao/examples/torch_custom_op/libtorch_custom_op.dylib"
)


def quantize(vals: torch.Tensor, group_size: int, nbit: int, scale_only: bool):
    assert nbit >= 2 and nbit <= 8
    qmin = -(1 << (nbit - 1))
    qmax = (1 << (nbit - 1)) - 1

    n, k = vals.shape
    vals = vals.reshape(-1, group_size)
    vmins, _ = torch.min(vals, axis=1)
    vmaxs, _ = torch.max(vals, axis=1)
    group_scales = (vmaxs - vmins) / (qmax - qmin)

    if scale_only:
        group_qvals = torch.round(vals / group_scales.reshape(-1, 1))
    else:
        group_zeros = qmin - torch.round(vmins / group_scales)
        group_qvals = torch.round(
            group_zeros.reshape(-1, 1) + vals / group_scales.reshape(-1, 1)
        )

    group_qvals = torch.clip(group_qvals, qmin, qmax).reshape(n, k).to(torch.int8)

    if scale_only:
        return group_qvals, group_scales
    return group_qvals, group_scales, group_zeros


def linear_a8sz_w_lowbit_reference_impl(
    weights, activations, group_size, nbit, has_weight_zeros
):
    n, k = weights.shape
    m, k = activations.shape
    assert m == 1
    assert k % group_size == 0

    if has_weight_zeros:
        weight_qvals, weight_scales, weight_zeros = quantize(
            weights, group_size, nbit, scale_only=False
        )
        weights_dequantized = (
            weight_scales.reshape(-1, 1)
            * (weight_qvals.reshape(-1, group_size) - weight_zeros.reshape(-1, 1))
        ).reshape(n, k)
    else:
        weight_qvals, weight_scales = quantize(
            weights, group_size, nbit, scale_only=True
        )
        weights_dequantized = (
            weight_scales.reshape(-1, 1) * (weight_qvals.reshape(-1, group_size))
        ).reshape(n, k)

    activation_qvals, activations_scales, activations_zeros = quantize(
        activations, k, 8, False
    )
    activations_dequantized = activations_scales * (
        activation_qvals - activations_zeros
    ).reshape(m, k)
    return torch.matmul(activations_dequantized, weights_dequantized.transpose(1, 0))


class _quantized_linear(nn.Module):
    def __init__(
        self,
        nbit,
        has_weight_zeros,
        pack_weight_op,
        linear_op,
        squeeze_unsqueeze_dim0=False,
    ):
        super().__init__()
        self.squeeze_unsqueeze_dim0 = squeeze_unsqueeze_dim0
        self.nbit = nbit

        self._has_weight_zeros = has_weight_zeros
        self._pack_weights_op = pack_weight_op
        self._linear_op = linear_op

    def pack_weights(self, weight_qvals, weight_scales_and_zeros, group_size):
        n, k = weight_qvals.shape

        # TODO(T200095131): convert self.n, self.k, self.group_size to
        # int when supported by AOTI
        self.n = torch.empty(n)
        self.k = torch.empty(k)
        self.group_size = torch.empty(group_size)

        if self._has_weight_zeros:
            weight_scales, weight_zeros = weight_scales_and_zeros
            self.packed_weights = self._pack_weights_op(
                weight_qvals, weight_scales, weight_zeros, self.group_size
            )
        else:
            weight_scales = weight_scales_and_zeros
            self.packed_weights = self._pack_weights_op(
                weight_qvals, weight_scales, self.group_size
            )

    def forward(self, x):
        if self.squeeze_unsqueeze_dim0:
            x = x.squeeze(0)

        res = self._linear_op(self.packed_weights, self.n, self.k, self.group_size, x)

        if self.squeeze_unsqueeze_dim0:
            res = res.unsqueeze(0)
        return res


def replace_linear_with_quantized_linear(module: nn.Module, kwargs={}):
    group_size = kwargs["group_size"]
    nbit = kwargs["nbit"]
    has_weight_zeros = kwargs["has_weight_zeros"]
    squeeze_unsqueeze_dim0 = (
        kwargs["squeeze_unsqueeze_dim0"]
        if "squeeze_unsqueeze_dim0" in kwargs
        else False
    )

    for name, child in module.named_children():
        if isinstance(child, nn.Linear):
            assert child.bias is None

            if not has_weight_zeros:
                weight_qvals, weight_scales = quantize(
                    child.weight, group_size=group_size, nbit=nbit, scale_only=True
                )
                weight_scales_and_zeros = weight_scales
            else:
                weight_qvals, weight_scales, weight_zeros = quantize(
                    child.weight, group_size=group_size, nbit=nbit, scale_only=False
                )
                weight_scales_and_zeros = (weight_scales, weight_zeros.to(torch.int8))

            qlinear = None
            if nbit == 2:
                if has_weight_zeros:
                    qlinear = _quantized_linear(
                        nbit=nbit,
                        has_weight_zeros=has_weight_zeros,
                        pack_weight_op=torch.ops.torchao._pack_weights_a8sz_w2sz,
                        linear_op=torch.ops.torchao._linear_a8sz_w2sz,
                        squeeze_unsqueeze_dim0=squeeze_unsqueeze_dim0,
                    )
                else:
                    qlinear = _quantized_linear(
                        nbit=nbit,
                        has_weight_zeros=has_weight_zeros,
                        pack_weight_op=torch.ops.torchao._pack_weights_a8sz_w2s,
                        linear_op=torch.ops.torchao._linear_a8sz_w2s,
                        squeeze_unsqueeze_dim0=squeeze_unsqueeze_dim0,
                    )
            elif nbit == 3:
                if has_weight_zeros:
                    qlinear = _quantized_linear(
                        nbit=nbit,
                        has_weight_zeros=has_weight_zeros,
                        pack_weight_op=torch.ops.torchao._pack_weights_a8sz_w3sz,
                        linear_op=torch.ops.torchao._linear_a8sz_w3sz,
                        squeeze_unsqueeze_dim0=squeeze_unsqueeze_dim0,
                    )
                else:
                    qlinear = _quantized_linear(
                        nbit=nbit,
                        has_weight_zeros=has_weight_zeros,
                        pack_weight_op=torch.ops.torchao._pack_weights_a8sz_w3s,
                        linear_op=torch.ops.torchao._linear_a8sz_w3s,
                        squeeze_unsqueeze_dim0=squeeze_unsqueeze_dim0,
                    )
            elif nbit == 4:
                if has_weight_zeros:
                    qlinear = _quantized_linear(
                        nbit=nbit,
                        has_weight_zeros=has_weight_zeros,
                        pack_weight_op=torch.ops.torchao._pack_weights_a8sz_w4sz,
                        linear_op=torch.ops.torchao._linear_a8sz_w4sz,
                        squeeze_unsqueeze_dim0=squeeze_unsqueeze_dim0,
                    )
                else:
                    qlinear = _quantized_linear(
                        nbit=nbit,
                        has_weight_zeros=has_weight_zeros,
                        pack_weight_op=torch.ops.torchao._pack_weights_a8sz_w4s,
                        linear_op=torch.ops.torchao._linear_a8sz_w4s,
                        squeeze_unsqueeze_dim0=squeeze_unsqueeze_dim0,
                    )
            elif nbit == 5:
                if has_weight_zeros:
                    qlinear = _quantized_linear(
                        nbit=nbit,
                        has_weight_zeros=has_weight_zeros,
                        pack_weight_op=torch.ops.torchao._pack_weights_a8sz_w5sz,
                        linear_op=torch.ops.torchao._linear_a8sz_w5sz,
                        squeeze_unsqueeze_dim0=squeeze_unsqueeze_dim0,
                    )
                else:
                    qlinear = _quantized_linear(
                        nbit=nbit,
                        has_weight_zeros=has_weight_zeros,
                        pack_weight_op=torch.ops.torchao._pack_weights_a8sz_w5s,
                        linear_op=torch.ops.torchao._linear_a8sz_w5s,
                        squeeze_unsqueeze_dim0=squeeze_unsqueeze_dim0,
                    )
            else:
                raise ValueError(
                    f"Unsupported nbit ({nbit}) and has_weight_zeros ({has_weight_zeros}) combination"
                )

            assert qlinear is not None
            setattr(module, name, qlinear)
            getattr(module, name).pack_weights(
                weight_qvals,
                weight_scales_and_zeros,
                group_size,
            )
        else:
            replace_linear_with_quantized_linear(child, kwargs)
